### ISSUE TITLE
[Feature Request] Generalize `beartype.door.{is_bearable,die_if_unbearable}` to accept generic types in forward refrences

### ISSUE THREAD

**Issue by Logan-Pageler** [beartype/beartype#414]
Currently `beartype.door.{is_bearable,die_if_unbearable}` doesn't seem to be able to accept forward references when they have a generic type. Consider the following:

```python
from collections.abc import Sequence
from typing import Generic, TypeVar

from beartype.door import is_bearable

T = TypeVar('T')
class Foo(Generic[T]):
  ...

class Bar(Sequence['__main__.Foo[float]']):  # if i remove "[float]" this entire code works
  ...

bar = Bar()
is_bearable(bar, Bar)  # Throws BeartypeDecorHintForwardRefException
```
yields the error:
```
beartype.roar.BeartypeDecorHintForwardRefException: Forward reference '__main__.Foo[float]' not valid Python attribute name.
```

I get this seems similar to #364 and I figure solving #364 would hopefully entail solving this, but if I understand correctly, thats pretty far away. And I just can't seem to find a good work around for this one. So other insight to how to solve this temporarily would also be greatly appreciated.

**Comment by leycec** [comment]
*...heh.* Forward reference resolution in `is_bearable()`, huh? Sounds great! I *definitely* should have done that... like, decades ago. There's actually no significant blocker to me doing that. Well, *okay*. The only significant blockers to me doing that are sanity and time – and I have copious amounts of both of those.

Technically, this is indeed a duplicate of #364. Please prod me if I fail to do this promptly for you and everybody else who clearly kinda desperately needs this. I didn't know this was in such hot demand! Thanks so much for the friendly heads-up. :+1: :smile_cat: :+1:

In the meanwhile, you *might* be able to temporarily circumvent this deficit in @beartype with a *horrible* `typing.TYPE_CHECKING` hack like:

```python
from typing import TYPE_CHECKING, Generic

# If this module is currently being statically type-checked,
# prefer a full-blown subscripted generic forward reference.
# Sadly, is_bearable() currently fails to support this. Sad!
if TYPE_CHECKING:
    _BarSuperclass = Sequence['__main__.Foo[float]']
# Else, this module is being interpreted by a Python interpreter.
# In this case, reduce this to an unsubscripted generic forward
# reference that @beartype can actually handle.
else:
    _BarSuperclass = Sequence['__main__.Foo']

class Bar(Generic[float], _BarSuperclass):
    ...
```

Does that actually work? Probably not. Looks kinda sketchy, huh? I'm squinting at that and not liking what I'm seeing either.

Let's quietly close this in favour of #364, which I really should just do already. :face_exhaling:

**Comment by leycec** [comment]
**WAIT.** Technically, [the horrible hack](https://github.com/beartype/beartype/issues/414#issuecomment-2288024153) above *does* technically work. That's nice. Interestingly, however, this issue actually has *nothing* to do with #364. This is a novel (and fairly serious) low-level @beartype bug that nobody has reported before. That's... *not so nice.*

Most @beartype users (and Pythonistas in general, really) ignore subscripted generics like `typing.Generic[T]`. This is why you're the first to sadly faceplant into this issue. Let's reopen this under a brand new title. Onward to issue resolution! :fist_raised:

**Comment by Logan-Pageler** [comment]
Yeah, thanks for suggested fix. While I agree its a bit messy, it should be fine for what I'm doing for now so thanks. But, I definitely appreciate treating this separately from #364 as this bug is more breaking in my eyes. I debated for a while whether to call it a bug or a feature request lol.

**Comment by leycec** [comment]
> ...as this bug is more breaking in my eyes.

**Totally.** You got it right. There is breakage here and this saddens me. I'm currently obsessed with an unrelated feature request #414. Just as soon as I finish that, I'll switch on over to this and resolve this for our upcoming @beartype 0.19.0 release bonanza. This is critical stuff.

Thanks again, @Logan-Pageler! You're the Wolverine of QA. :wolf:

**Comment by leycec** [comment]
**Resolved by 010af0639650ce.** 大成功!!!! Massive victory!!! Huge success!! @beartype invokes thee, Coffee Cup Techlead!

![](https://media1.tenor.com/m/GEfBX3DNubEAAAAd/tired-of-being-so-successful-techlead.gif)
<sup>*no. @beartype never tires of this.*</sup>

Our upcoming @beartype `0.23.0` is The Big Forward Reference Fix the World Has Been Waiting For. @beartype `0.23.0` resolves this and all other open feature requests concerning forward references. This magic could not have been possible without high-functioning autism and the Canadian winter. Together, the two combined to produce an almost sickening attention to obsessive-compulsive detail and code minutiae no other living human could possibly have cared above.

I refactored the *entire* @beartype codebase. Hundreds of thousands of lines of BadSadCode™ were "disappeared" in favour of hundreds of thousands of lines of GoodHappyCode™. It shouldn't have worked. I almost gave up at multiple junctures. I was *this* close to scrapping the whole `bigglyforwardrefhackjob` feature branch.

But autism and Canada saw us through. What else could anyone do? It's literally -32°C out this window. The snow drifts and icicles the size of my entire body were almost the end of every Canadian multiple times. Yet, coding intensified. This is its work product. Obsession personified. @beartype `0.23.0`. The end of the line and final word on forward references.

In short, forward references just "work" now. They didn't before. @beartype only supported a few specific common use cases for forward references... while unsupporting all possible use cases for forward references. The monsters of the past are now dead.

@beartype `0.23.0` features a full-blown forward referencing engine. It just magically works – always, everywhere, for all use cases. Therefore, I implore everyone to...

## Prove @leycec Wrong: It Can't Be *That* Hard

In 2026, does anyone still care about low-level QA hardships like runtime forward reference resolution? No? Just me, huh? I'm all that's left. I'm the last of us. Humanity has long-since moved on to cool stuff like sentient machines, ~~murderous~~ "defensive" robots with paradoxically gentle emotions, and asteroid mining. Great. That's just great.

Would those who are still here mind testing these immense changes out on your own favourite codebase? If so, we beg you to:

```bash
pip install --upgrade git+https://github.com/beartype/beartype.git@010af0639650ce106b08136541c16e4c995e3105
```

Fleshy meat sacks like me can only predict so many real-world friction points. Edge cases and fanged dragons abound. The @beartype test suite is ponderous in its girth. Yet, failure is still guaranteed.

Help @beartype help you. This is how we QA. We fail together as one or not at all! 💪🐻  <sup>*<-- maybe not the best motivational one-liner*</sup>
